File: dictator-kotlin/dictator-core/src/commonMain/kotlin/com/dictator/core/data/privacy/ProviderPolicyManager.kt
/**
 * Provider Policy Manager
 * Manages AI provider privacy policies and data handling practices
 */
package com.dictator.core.data.privacy

/**
 * Manager for AI provider privacy policies
 */
class ProviderPolicyManager {
    
    private val policies = mutableMapOf<String, AiProviderPolicy>()
    
    init {
        // Initialize with default policies
        initializeDefaultPolicies()
    }

    /**
     * Initialize default provider policies
     */
    private fun initializeDefaultPolicies() {
        // Anthropic Claude Policy
        policies["claude"] = AiProviderPolicy(
            id = "policy-claude-v1",
            provider = "claude",
            displayName = "Anthropic Claude",
            dataRetentionDays = 30,
            processingPurposes = listOf("service-improvement", "user-support", "security"),
            processingLocations = listOf("us"),
            usesDataForTraining = false,
            trainingOptOutAvailable = true,
            privacyPolicyUrl = "https://www.anthropic.com/privacy",
            gdprCompliant = true,
            notes = "User conversations are deleted after 30 days by default. Available via Claude API."
        )

        // OpenAI Policy
        policies["openai"] = AiProviderPolicy(
            id = "policy-openai-v1",
            provider = "openai",
            displayName = "OpenAI",
            dataRetentionDays = 30,
            processingPurposes = listOf("model-training", "service-improvement", "user-support"),
            processingLocations = listOf("us"),
            usesDataForTraining = true,
            trainingOptOutAvailable = true,
            privacyPolicyUrl = "https://openai.com/privacy",
            gdprCompliant = true,
            notes = "Data may be used for model training unless opted out."
        )

        // Google Policy
        policies["google"] = AiProviderPolicy(
            id = "policy-google-v1",
            provider = "google",
            displayName = "Google",
            dataRetentionDays = null, // Indefinite
            processingPurposes = listOf("model-training", "service-improvement"),
            processingLocations = listOf("us", "eu"),
            usesDataForTraining = true,
            trainingOptOutAvailable = true,
            privacyPolicyUrl = "https://policies.google.com/privacy",
            gdprCompliant = true,
            notes = "Retention policy varies by service. Opt-out available for training use."
        )

        // Local Ollama Policy
        policies["ollama"] = AiProviderPolicy(
            id = "policy-ollama-v1",
            provider = "ollama",
            displayName = "Ollama (Local)",
            dataRetentionDays = 0, // No retention - local processing
            processingPurposes = emptyList(),
            processingLocations = listOf("local"),
            usesDataForTraining = false,
            trainingOptOutAvailable = false,
            privacyPolicyUrl = "https://ollama.ai",
            gdprCompliant = true,
            notes = "100% local processing. No data transmission or retention. Highest privacy."
        )

        // Dictator's Policy (own backend)
        policies["dictator"] = AiProviderPolicy(
            id = "policy-dictator-v1",
            provider = "dictator",
            displayName = "Dictator Server",
            dataRetentionDays = 90,
            processingPurposes = listOf("service-improvement", "user-support", "security"),
            processingLocations = listOf("us"),
            usesDataForTraining = false,
            trainingOptOutAvailable = false,
            privacyPolicyUrl = "https://www.getdictator.ai/privacy",
            gdprCompliant = true,
            notes = "Data encrypted in transit and at rest. Retained for 90 days for audit."
        )
    }

    /**
     * Get policy for a specific provider
     */
    fun getPolicy(provider: String): AiProviderPolicy? {
        return policies[provider.lowercase()]
    }

    /**
     * Get all registered provider policies
     */
    fun getAllPolicies(): List<AiProviderPolicy> {
        return policies.values.toList()
    }

    /**
     * Add or update a provider policy
     */
    fun registerPolicy(policy: AiProviderPolicy) {
        policies[policy.provider.lowercase()] = policy
    }

    /**
     * Remove a provider policy
     */
    fun removePolicy(provider: String) {
        policies.remove(provider.lowercase())
    }

    /**
     * Check if provider is GDPR compliant
     */
    fun isGdprCompliant(provider: String): Boolean {
        return getPolicy(provider)?.gdprCompliant ?: false
    }

    /**
     * Check if provider uses data for training
     */
    fun usesDataForTraining(provider: String): Boolean {
        return getPolicy(provider)?.usesDataForTraining ?: false
    }

    /**
     * Check if training opt-out is available
     */
    fun isTrainingOptOutAvailable(provider: String): Boolean {
        return getPolicy(provider)?.trainingOptOutAvailable ?: false
    }

    /**
     * Get data retention days
     */
    fun getDataRetentionDays(provider: String): Int? {
        return getPolicy(provider)?.dataRetentionDays
    }

    /**
     * Get processing purposes
     */
    fun getProcessingPurposes(provider: String): List<String> {
        return getPolicy(provider)?.processingPurposes ?: emptyList()
    }

    /**
     * Get processing locations
     */
    fun getProcessingLocations(provider: String): List<String> {
        return getPolicy(provider)?.processingLocations ?: emptyList()
    }

    /**
     * Get privacy policy URL
     */
    fun getPrivacyPolicyUrl(provider: String): String? {
        return getPolicy(provider)?.privacyPolicyUrl
    }

    /**
     * Check if provider is local (no data transmission)
     */
    fun isLocalProvider(provider: String): Boolean {
        val locations = getProcessingLocations(provider)
        return locations.contains("local")
    }

    /**
     * Rank providers by privacy (0.0 to 1.0, higher = more private)
     */
    fun getPrivacyScore(provider: String): Float {
        val policy = getPolicy(provider) ?: return 0.5f // Default middle score

        var score = 0.0f

        // Local processing = highest privacy (1.0)
        if (isLocalProvider(provider)) {
            return 1.0f
        }

        // GDPR compliance (0.2 points)
        if (policy.gdprCompliant) score += 0.2f

        // No data training (0.3 points)
        if (!policy.usesDataForTraining) score += 0.3f

        // Short retention (0.2 points)
        if (policy.dataRetentionDays != null && policy.dataRetentionDays <= 30) {
            score += 0.2f
        }

        // Limited processing purposes (0.15 points)
        if (policy.processingPurposes.size <= 2) score += 0.15f

        // US/EU only (0.1 points)
        val allUsEuLike = policy.processingLocations.all { 
            it in listOf("us", "eu", "uk", "ca", "au")
        }
        if (allUsEuLike) score += 0.1f

        return score.coerceIn(0.0f, 1.0f)
    }

    /**
     * Get providers sorted by privacy score (highest first)
     */
    fun getProvidersByPrivacy(): List<Pair<String, Float>> {
        return policies.map { (provider, _) ->
            provider to getPrivacyScore(provider)
        }.sortedByDescending { it.second }
    }

    /**
     * Compare two providers
     */
    fun compareProviders(provider1: String, provider2: String): String {
        val policy1 = getPolicy(provider1)
        val policy2 = getPolicy(provider2)
        
        if (policy1 == null || policy2 == null) {
            return "Provider not found"
        }

        val score1 = getPrivacyScore(provider1)
        val score2 = getPrivacyScore(provider2)
        val morePrivate = if (score1 > score2) provider1 else provider2
        val scoreFormat = "%.2f vs %.2f".format(score1, score2)

        return "$morePrivate is more private ($scoreFormat)"
    }

    /**
     * Get recommended provider based on privacy requirements
     */
    fun getRecommendedProvider(requireGdprCompliance: Boolean = true, allowTraining: Boolean = false): String? {
        val providers = getAllPolicies()
            .filter { 
                (!requireGdprCompliance || it.gdprCompliant) &&
                (allowTraining || !it.usesDataForTraining)
            }
            .sortedByDescending { getPrivacyScore(it.provider) }

        return providers.firstOrNull()?.provider
    }

    /**
     * Generate privacy report for a provider
     */
    fun generatePrivacyReport(provider: String): String {
        val policy = getPolicy(provider) ?: return "Provider not found"
        val score = getPrivacyScore(provider)

        val sb = StringBuilder()
        sb.append("=== Privacy Report: ${policy.displayName} ===\n")
        sb.append("Privacy Score: ${"%.1f".format(score * 100)}%\n")
        sb.append("GDPR Compliant: ${policy.gdprCompliant}\n")
        sb.append("Data for Training: ${policy.usesDataForTraining}\n")
        sb.append("Training Opt-Out: ${policy.trainingOptOutAvailable}\n")
        sb.append("Data Retention: ${policy.dataRetentionDays?.toString() ?: "Indefinite"} days\n")
        sb.append("Processing Locations: ${policy.processingLocations.joinToString(", ")}\n")
        sb.append("Processing Purposes: ${policy.processingPurposes.joinToString(", ")}\n")
        sb.append("Policy URL: ${policy.privacyPolicyUrl}\n")
        sb.append("Notes: ${policy.notes}\n")

        return sb.toString()
    }
}

File: lib/privacy/ProviderPolicyManager.ts
/**
 * AI Provider Policy Manager
 * Manages privacy policies for AI providers and tracks which policy is active
 * for each AI request
 */

import type { AiProviderPolicy } from './types';

/**
 * Default privacy policies for well-known providers
 * These should be updated as provider policies change
 */
const DEFAULT_PROVIDER_POLICIES: Record<string, Omit<AiProviderPolicy, 'id' | 'createdAt' | 'updatedAt'>> = {
  claude: {
    provider: 'claude',
    displayName: 'Anthropic Claude',
    dataRetentionDays: 30,
    processingPurposes: ['service-improvement'],
    processingLocations: ['us'],
    usesDataForTraining: false,
    trainingOptOutAvailable: true,
    privacyPolicyUrl: 'https://www.anthropic.com/privacy',
    gdprCompliant: true,
    notes: 'Claude API provides privacy-conscious handling of data. Conversations are generally not used for training.',
  },

  openai: {
    provider: 'openai',
    displayName: 'OpenAI',
    dataRetentionDays: 30,
    processingPurposes: ['service-improvement', 'model-training'],
    processingLocations: ['us'],
    usesDataForTraining: true,
    trainingOptOutAvailable: true,
    privacyPolicyUrl: 'https://openai.com/privacy',
    gdprCompliant: true,
    notes: 'OpenAI uses API data for model improvement. You can opt out for your organization.',
  },

  ollama: {
    provider: 'ollama',
    displayName: 'Ollama (Local)',
    dataRetentionDays: null,
    processingPurposes: [],
    processingLocations: ['on-device'],
    usesDataForTraining: false,
    trainingOptOutAvailable: false,
    privacyPolicyUrl: '',
    gdprCompliant: true,
    notes: 'Ollama runs locally on your device. Data never leaves your machine.',
  },

  dictator: {
    provider: 'dictator',
    displayName: 'Dictator (Local Inference)',
    dataRetentionDays: null,
    processingPurposes: [],
    processingLocations: ['on-device'],
    usesDataForTraining: false,
    trainingOptOutAvailable: false,
    privacyPolicyUrl: '',
    gdprCompliant: true,
    notes: 'Dictator runs on your device. No data is transmitted to external services.',
  },

  'openai-compatible': {
    provider: 'openai-compatible',
    displayName: 'OpenAI-Compatible API',
    dataRetentionDays: null,
    processingPurposes: ['service-improvement'],
    processingLocations: ['other'],
    usesDataForTraining: false,
    trainingOptOutAvailable: false,
    privacyPolicyUrl: 'https://example.com/privacy',
    gdprCompliant: false,
    notes: 'Privacy policy depends on the specific API endpoint you configure.',
  },
};

/**
 * AI Provider Policy Manager
 */
export class ProviderPolicyManager {
  private policies: Map<string, AiProviderPolicy> = new Map();
  private activePolicy: AiProviderPolicy | null = null;

  constructor(customPolicies?: AiProviderPolicy[]) {
    this.initializeDefaultPolicies();
    if (customPolicies) {
      for (const policy of customPolicies) {
        this.policies.set(policy.id, policy);
      }
    }
  }

  /**
   * Initialize default policies from built-in definitions
   */
  private initializeDefaultPolicies(): void {
    const now = Date.now();
    for (const [provider, config] of Object.entries(DEFAULT_PROVIDER_POLICIES)) {
      const policy: AiProviderPolicy = {
        id: `policy-${provider}-${Math.floor(now / 1000)}`,
        createdAt: now,
        updatedAt: now,
        ...config,
      };
      this.policies.set(policy.id, policy);
    }
  }

  /**
   * Get policy by ID
   */
  getPolicyById(policyId: string): AiProviderPolicy | null {
    return this.policies.get(policyId) || null;
  }

  /**
   * Get latest policy for a provider
   */
  getLatestPolicyForProvider(provider: string): AiProviderPolicy | null {
    let latest: AiProviderPolicy | null = null;
    for (const policy of this.policies.values()) {
      if (policy.provider === provider) {
        if (!latest || policy.updatedAt > latest.updatedAt) {
          latest = policy;
        }
      }
    }
    return latest;
  }

  /**
   * Get all policies for a provider (versioned)
   */
  getPoliciesForProvider(provider: string): AiProviderPolicy[] {
    const results: AiProviderPolicy[] = [];
    for (const policy of this.policies.values()) {
      if (policy.provider === provider) {
        results.push(policy);
      }
    }
    return results.sort((a, b) => b.updatedAt - a.updatedAt);
  }

  /**
   * Get all available providers
   */
  getAvailableProviders(): Array<{ provider: string; displayName: string }> {
    const providers = new Map<string, string>();
    for (const policy of this.policies.values()) {
      if (!providers.has(policy.provider)) {
        providers.set(policy.provider, policy.displayName);
      }
    }
    return Array.from(providers.entries()).map(([provider, displayName]) => ({
      provider,
      displayName,
    }));
  }

  /**
   * Add or update a policy
   */
  upsertPolicy(policy: AiProviderPolicy): void {
    this.policies.set(policy.id, policy);
  }

  /**
   * Get privacy rating for a provider (0-100)
   * Higher = more privacy-friendly
   */
  getPrivacyRating(provider: string): number {
    const policy = this.getLatestPolicyForProvider(provider);
    if (!policy) {
      return 0;
    }

    let rating = 50; // Base score

    // Local processing gets high score
    if (policy.processingLocations.includes('on-device')) {
      rating = 95;
    } else {
      // Non-training data usage
      if (!policy.usesDataForTraining) {
        rating += 30;
      } else if (policy.trainingOptOutAvailable) {
        rating += 20;
      }

      // GDPR compliance
      if (policy.gdprCompliant) {
        rating += 15;
      }

      // Data retention
      if (policy.dataRetentionDays !== null) {
        if (policy.dataRetentionDays <= 30) {
          rating += 10;
        }
      }
    }

    return Math.min(100, Math.max(0, rating));
  }

  /**
   * Get privacy summary for a provider (human-readable)
   */
  getPrivacySummary(provider: string): string {
    const policy = this.getLatestPolicyForProvider(provider);
    if (!policy) {
      return 'Policy information not available';
    }

    const parts: string[] = [];

    if (policy.processingLocations.includes('on-device')) {
      parts.push('✅ Processes data locally on your device');
    } else {
      parts.push(`📍 Processes data in: ${policy.processingLocations.join(', ')}`);
    }

    if (!policy.usesDataForTraining) {
      parts.push('✅ Does not use data for model training');
    } else if (policy.trainingOptOutAvailable) {
      parts.push('⚠️  Uses data for training (but you can opt-out)');
    } else {
      parts.push('⚠️  Uses data for model training');
    }

    if (policy.dataRetentionDays !== null) {
      parts.push(`🗑️  Retains data for ${policy.dataRetentionDays} days`);
    } else if (!policy.processingLocations.includes('on-device')) {
      parts.push('📦 Data retention period not specified');
    }

    if (policy.gdprCompliant) {
      parts.push('✅ GDPR compliant');
    }

    return parts.join('\n');
  }

  /**
   * Validate if a provider can be used for a specific use case
   */
  validateForUseCase(
    provider: string,
    requiresTrainingOptOut: boolean = false,
    requiresGdprCompliance: boolean = false
  ): { valid: boolean; reason?: string } {
    const policy = this.getLatestPolicyForProvider(provider);
    if (!policy) {
      return { valid: false, reason: 'Provider policy not found' };
    }

    if (requiresTrainingOptOut && policy.usesDataForTraining && !policy.trainingOptOutAvailable) {
      return { valid: false, reason: 'Provider uses data for training and does not support opt-out' };
    }

    if (requiresGdprCompliance && !policy.gdprCompliant) {
      return { valid: false, reason: 'Provider is not GDPR compliant' };
    }

    return { valid: true };
  }

  /**
   * Get recommendations for best privacy practices
   */
  getPrivacyRecommendations(): string[] {
    const recommendations: string[] = [];

    // Check if any local providers are available
    const localProviders = this.getAvailableProviders().filter(
      (p) => this.getLatestPolicyForProvider(p.provider)?.processingLocations.includes('on-device')
    );

    if (localProviders.length > 0) {
      recommendations.push(
        `💡 For maximum privacy, consider using local providers: ${localProviders.map((p) => p.displayName).join(', ')}`
      );
    }

    // Check for training-avoiding providers
    const noTrainingProviders = this.getAvailableProviders().filter((p) => {
      const policy = this.getLatestPolicyForProvider(p.provider);
      return policy && !policy.usesDataForTraining;
    });

    if (noTrainingProviders.length > 0) {
      recommendations.push(
        `💡 Providers that don't use data for training: ${noTrainingProviders.map((p) => p.displayName).join(', ')}`
      );
    }

    // General recommendations
    recommendations.push('💡 Review AI data policies in Settings before sending sensitive content');
    recommendations.push('💡 Use "Selected Text Only" mode when possible instead of full document');
    recommendations.push('💡 Enable sensitive data detection warnings');

    return recommendations;
  }
}

/**
 * Create default policy manager with built-in policies
 */
export function createDefaultPolicyManager(): ProviderPolicyManager {
  return new ProviderPolicyManager();
}

/**
 * Global singleton instance
 */
let globalPolicyManager: ProviderPolicyManager | null = null;

/**
 * Initialize global policy manager
 */
export function initializePolicyManager(customPolicies?: AiProviderPolicy[]): ProviderPolicyManager {
  globalPolicyManager = new ProviderPolicyManager(customPolicies);
  return globalPolicyManager;
}

/**
 * Get global policy manager instance
 */
export function getPolicyManager(): ProviderPolicyManager {
  if (!globalPolicyManager) {
    globalPolicyManager = createDefaultPolicyManager();
  }
  return globalPolicyManager;
}
